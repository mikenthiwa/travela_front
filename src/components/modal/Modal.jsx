import React, { PureComponent, Fragment } from 'react';
import { PropTypes } from 'prop-types';
import Overlay from './overlay/Overlay';
import closeButton from '../../images/icons/close.svg';
import './_modal.scss';

class Modal extends PureComponent {
  _isMounted = false;
  state = {
    showing: false
  };

  componentDidMount() {
    this._isMounted = true;
    document.addEventListener('keydown', this.hideModal);
  }

  componentWillReceiveProps( { visibility }, nextContext) {
    this.timeout = setTimeout(() => {
      this.setState({ showing: visibility === 'visible'});
    }, visibility === 'visible' ? 100 : 200);
  }

  componentWillUnmount() {
    this._isMounted = false;
    document.removeEventListener('keydown', this.hideModal);
    if( this.timeout ){
      clearTimeout(this.timeout);
    }
  }

  hideModal = (e) => {
    const { closeModal } = this.props;
    e.key === 'Escape' && closeModal && closeModal();
  };

  renderModalHeader = () => {
    const { title, closeModal, modalBar, closeDeleteModal, requestId } = this.props;
    let url = location.pathname;
    if (requestId) {
      let urlArr = url.split('/');
      url = urlArr.slice(0, urlArr.length - 1).join('/');
    }
    if (/requests\/\w+\/checklist/.test(url)) {
      url = '/requests';
    }
    return (
      <div className="modal-title-bar">
        <div className="modal-title-text">
          <div className="modal-title modal-title-word">
            {title}
          </div>
          <div className="lable-text modal-bar">
            {modalBar}
          </div>
        </div>
        {closeDeleteModal ?
          (
            <button type="button" onClick={closeDeleteModal} className="modal-close">
              <img alt="close" src={closeButton} />
            </button>
          )
          :
          (
            (url.includes('/requests')) ?
              (
                <button
                  type="button" onClick={closeDeleteModal ?
                    closeDeleteModal : closeModal} className="modal-close">
                  <img alt="close" src={closeButton} />
                </button>
              ) : (
                <button
                  type="button" onClick={closeDeleteModal ?
                    closeDeleteModal : closeModal} className="modal-close">
                  <img alt="close" src={closeButton} />
                </button>
              )
          )
        }
      </div>
    );
  };


  renderContent = (
    showing, visibility, customModalStyles,
    width, modalId, modalContentId, children
  ) => (
    <div
      className={`modal ${ showing ? visibility : 'invisible' } ${customModalStyles}`}
      style={{maxWidth: width}}
      onClick={e => {e.stopPropagation();}} onKeyPress={() => {}}
      id={modalId}
      tabIndex="0"
      role="button">
      {this.renderModalHeader()}
      <div className="modal-content" id={modalContentId}>
        {children}
      </div>
    </div>
  );

  render() {
    const {
      children, visibility, width, modalId,
      modalContentId, showOverlay, customOverlayStyle,
      customModalStyles, closeModal
    } = this.props;
    const { showing } = this.state;
    const overlayStyle = `${ showing ? visibility : 'invisible' } ${customOverlayStyle}`;

    const showModal = showing || visibility === 'visible';
    return showModal ? (
      <Fragment>
        <Overlay
          closeModal={closeModal}
          className={`${overlayStyle}`}
          overlayBackground={!showOverlay ? 'overlayBackground' : ''}>
          {this.renderContent(
            showing, visibility, customModalStyles,
            width, modalId, modalContentId, children)
          }
        </Overlay>
      </Fragment>
    ): null;
  }
}


Modal.propTypes = {
  visibility: PropTypes.oneOf(['visible', 'invisible']).isRequired,
  closeModal: PropTypes.func,
  title: PropTypes.string,
  modalId: PropTypes.string,
  modalContentId: PropTypes.string,
  width:  PropTypes.string,
  showOverlay: PropTypes.bool,
  modalBar: PropTypes.object,
  customModalStyles: PropTypes.string,
  closeDeleteModal: PropTypes.func,
  requestId: PropTypes.string,
  customOverlayStyle: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.shape({}),
    PropTypes.array
  ]),
};

Modal.defaultProps = {
  title: '',
  modalId: '',
  requestId: '',
  width: '',
  modalContentId: '',
  customOverlayStyle: '',
  customModalStyles: '',
  showOverlay: true,
  modalBar: <div />,
  closeModal: null,
  closeDeleteModal: null,
  children: {},
};

export default Modal;
