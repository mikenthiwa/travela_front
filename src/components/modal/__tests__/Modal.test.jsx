import React from 'react';
import Modal from '../Modal';

const EventQueue = function() {
  this.events = {};
  this.add = (event, callback) => {
    this.events[event] = [...this.getListeners(event), callback];
  };

  this.getListeners = (event) => {
    return this.events[event] || [];
  };

  this.removeListener = (event, callback) => {
    const list = this.getListeners(event);
    const index = list.indexOf(callback);
    if( index > 0){
      list.splice(index, 1);
    }
    this.events[event] = list;
  };

  this.fire = (type, event) => {
    this.getListeners(type).forEach(callback => callback(event));
  };
};

const eventQueue = new EventQueue();
document.addEventListener = jest.fn((event, callback) => {
  eventQueue.add(event, callback);
});

document.removeEventListener = jest.fn((event, callback) => {
  eventQueue.removeListener(event, callback);
});

const props = (overrides = {}) => ({
  title: 'test modal',
  visibility: 'visible',
  closeModal: jest.fn(),
  closeDeleteModal: jest.fn(),
  requestId: '34uoj31OJ',
  ...overrides
});
describe('<Modal />', () => {
  let click, wrapper, currentProps = props();

  beforeEach(() => {
    click = jest.fn();
    wrapper = mount(
      <Modal {...currentProps}>
        <div>
          Test content
        </div>
      </Modal>
    );
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('prevents closing modal on clicking modal body', () => {
    const stopPropagation = jest.fn();
    const modal = wrapper.find('div.modal').simulate('click', {
      stopPropagation
    });
    expect(stopPropagation).toHaveBeenCalled();
  });

  it('should hide modal on escape clicked', () => {
    const event = { key : 'Escape'};
    eventQueue.fire('keydown', event);

    expect(currentProps.closeModal).toHaveBeenCalled();
  });

  it('should release the keydown listener on unmount', () => {
    wrapper.unmount();
    expect(document.removeEventListener).toHaveBeenCalled();
  });

  describe('Modal Animations', () => {
    let callback;
    beforeEach(() => {
      wrapper = mount(
        <Modal {...props({visibility: 'invisible'})}>
          <div>
            Modal Content
          </div>
        </Modal>
      );
      jest.resetAllMocks();


      window.setTimeout = jest.fn((func) => {
        callback = func;
      });
    });

    it('should show a fade in animation', () => {
      wrapper.setProps({ visibility: 'visible'});

      expect(window.setTimeout.mock.calls[0][1]).toEqual(100);
      expect(wrapper.find('Overlay').props().className).toContain('invisible');
      expect(wrapper.find('div.modal').props().className).toContain('invisible');
      expect(wrapper.find('.modal-content').length).toEqual(1);

      callback();
      wrapper.update();

      expect(wrapper.find('Overlay').props().className).toEqual('visible ');
      expect(wrapper.find('div.modal').props().className).toContain('visible');
    });

    it('should show a fade out animation', () => {
      wrapper.setProps({ visibility: 'visible'});
      callback();

      wrapper.setProps({ visibility: 'invisible'});

      expect(window.setTimeout.mock.calls[1][1]).toEqual(200);
      expect(wrapper.find('Overlay').props().className).toContain('invisible');
      expect(wrapper.find('div.modal').props().className).toContain('invisible');
    });
  });
});
