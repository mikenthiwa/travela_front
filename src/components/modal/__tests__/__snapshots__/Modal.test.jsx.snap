// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders correctly 1`] = `
<Fragment>
  <Overlay
    className="visible "
    click={null}
    overlayBackground=""
  >
    <div
      className="modal visible "
      id=""
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "maxWidth": "",
        }
      }
      tabIndex="0"
    >
      <div
        className="modal-title-bar"
      >
        <div
          className="modal-title-text"
        >
          <div
            className="modal-title modal-title-word"
          >
            test modal
          </div>
          <div
            className="lable-text modal-bar"
          >
            <div />
          </div>
        </div>
        <button
          className="modal-close"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="close"
            src="close.svg"
          />
        </button>
      </div>
      <div
        className="modal-content"
        id=""
      >
        <div>
          Test content
        </div>
      </div>
    </div>
  </Overlay>
</Fragment>
`;
